File: 00_SHARE/01_past_questions/01_QE/22H1/22H1_Q1.py
## Q1 ##

### (a)
def palindrome(s: str) -> bool:
    flag = False
    if s == s[::-1]:
        flag = True
    return flag

### (b)
def substring(s: str, t: str) -> bool:
    flag = False
    if t in s:
        flag = True
    return flag

### (c)
def max_palindromes(s: str) -> list:

    #0. set parameters
    s_length = len(s)
    pal_list = []
    max_pals_list = []

    #1. find the distinct palindrome-substrings
    for i in range(s_length):
        for j in range(i+1, s_length+1):
            target = s[i:j]
            # print(f"({i}, {j}): {target}")            
            if (palindrome(target)) and (target not in pal_list):
                pal_list.append(target)

    #2. make max_palindrome list
    for i in pal_list:
        #2-1. set flag
        max_flag = True

        #2-2. check whether i is not same with j, but substring of j --> max_flag = False
        for j in pal_list:
            '''
            i: target string
            j: comparing string
            '''
            if i != j and substring(j, i):
                max_flag = False
                break
        
        #2-3. add if max_flag == True
        if max_flag:
            max_pals_list.append(i)

    return max_pals_list


# Driver Code
if __name__ == '__main__':
    testa1 = "asdsa"
    testa2 = "asddsa"
    testa3 = "asefavsae"

    testc1 = "kabccbadzdefgfeda"
    testc2 = "kabccba dzabccbaza"
    testc3 = "bbacaz"

    sub1 = 'a'
    sub2 = 'f'

    # (a) palindrome test
    print("\n(a) test")
    print(palindrome(testa1)) #True
    print(palindrome(testa2)) #True
    print(palindrome(testa3)) #False

    # (b) substring test
    print("\n(b) test")
    print(substring(testc1, sub1)) #True
    print(substring(testc2, sub1)) #True
    print(substring(testc1, sub2)) #True
    print(substring(testc2, sub2)) #False

    # (c) max_palindrome test
    print("\n(c) test")
    print(max_palindromes(testc1))
    print(max_palindromes(testc2))
    print(max_palindromes(testc3))

File: 00_SHARE/01_past_questions/01_QE/22H1/22H1_Q1.cpp
#include <iostream>
#include <string>
#include <functional>
#include <vector>

using namespace std;

bool palindrome(const string& s) {
    bool flag = false;
    if (s == string(s.rbegin(), s.rend())) {
        flag = true;
    }
    return flag;
}

bool substring(const string& s, const string& t) {
    bool flag = false;
    if (s.find(t) != string::npos) {
        flag = true;
    }
    return flag;
}

vector<string> max_palindromes(const string& s) {
    int s_length = s.length();
    vector<string> pal_list;
    vector<string> max_pals_list;

    for (int i = 0; i < s_length; i++) {
        for (int j = i + 1; j <= s_length; j++) {
            string target = s.substr(i, j - i);
            if (palindrome(target) && find(pal_list.begin(), pal_list.end(), target) == pal_list.end()) {
                pal_list.push_back(target);
            }
        }
    }

    for (const string& i : pal_list) {
        bool max_flag = true;
        for (const string& j : pal_list) {
            if (i != j && substring(j, i)) {
                max_flag = false;
                break;
            }
        }
        if (max_flag) {
            max_pals_list.push_back(i);
        }
    }

    return max_pals_list;
}

int main() {
    string testa1 = "asdsa";
    string testa2 = "asddsa";
    string testa3 = "asefavsae";

    string testc1 = "kabccbadzdefgfeda";
    string testc2 = "kabccba dzabccbaza";
    string testc3 = "bbacaz";

    string sub1 = "a";
    string sub2 = "f";

    // (a) palindrome test
    cout << "(a) test" << endl;
    cout << palindrome(testa1) << endl; // True
    cout << palindrome(testa2) << endl; // True
    cout << palindrome(testa3) << endl; // False

    // (b) substring test
    cout << "\n(b) test" << endl;
    cout << substring(testc1, sub1) << endl; // True
    cout << substring(testc2, sub1) << endl; // True
    cout << substring(testc1, sub2) << endl; // True
    cout << substring(testc2, sub2) << endl; // False

    // (c) max_palindrome test
    cout << "\n(c) test" << endl;

    cout << "----[1]----" << endl;
    vector<string> max_pals1 = max_palindromes(testc1);
    for (const string& max_pal : max_pals1) {
        cout << max_pal << endl;
    }
    cout << "-----------" << endl;

    cout << "----[2]----" << endl;
    vector<string> max_pals2 = max_palindromes(testc2);
    for (const string& max_pal : max_pals2) {
        cout << max_pal << endl;
    }
    cout << "-----------" << endl;

    cout << "----[3]----" << endl;
    vector<string> max_pals3 = max_palindromes(testc3);
    for (const string& max_pal : max_pals3) {
        cout << max_pal << endl;
    }
    cout << "-----------" << endl;

    return 0;
}
